grr/server/grr_response_server/databases/spanner_flows.py

```diff
@@ -0,0 +1,2512 @@
+      every incoming response.
+
+      currently_available_requests: a set with all the request ids corresponding
+      to given responses.
+    """
+
+    # Number of responses each affected request is waiting for (if available).
+    responses_expected_by_request = {}
+
+    # We also store all requests we have in the db so we can discard responses
+    # for unknown requests right away.
+    currently_available_requests = set()
+
+    # Callback states by request.
+    callback_state_by_request = {}
+
+    keys = []
+    for r in responses:
+      keys.append([r.client_id, r.flow_id, str(r.request_id)])
+
+    for row in txn.read(
+        table="FlowRequests",
+        keyset=spanner_lib.KeySet(keys=keys),
+        columns=[
+            "ClientID",
+            "FlowID",
+            "RequestID",
+            "CallbackState",
+            "ExpectedResponseCount",
+        ]
+    ):
+
+      request_key = _RequestKey(
+          row[0],
+          row[1],
+          int(row[2]),
+      )
+      currently_available_requests.add(request_key)
+
+      callback_state: str = row[3]
+      if callback_state:
+        callback_state_by_request[request_key] = callback_state
+
+      responses_expected: int = row[4]
+      if responses_expected:
+        responses_expected_by_request[request_key] = responses_expected
+
+    return (
+        responses_expected_by_request,
+        callback_state_by_request,
+        currently_available_requests,
+    )
+
+  def _BuildResponseWrites(
+      self,
+      responses: Collection[
+          Union[
+              flows_pb2.FlowResponse,
+              flows_pb2.FlowStatus,
+              flows_pb2.FlowIterator,
+          ],
+      ],
+      txn,
+  ) -> None:
+    """Builds the writes to store given responses in the db.
+
+    Args:
+      responses: iterable with flow responses to write.
+      txn: transaction to use for the writes.
+
+    Raises:
+      TypeError: if responses have objects other than FlowResponse, FlowStatus
+          or FlowIterator.
+    """
+    columns = ["ClientId",
+               "FlowId",
+               "RequestId",
+               "ResponseId",
+               "Response",
+               "Status",
+               "Iterator",
+               "CreationTime"]
+    rows = []
+    for r in responses:
+      response = None
+      status = None
+      iterator = None
+      if isinstance(r, flows_pb2.FlowResponse):
+        response = r
+      elif isinstance(r, flows_pb2.FlowStatus):
+        status = r
+      elif isinstance(r, flows_pb2.FlowIterator):
+        iterator = r
+      else:
+        # This can't really happen due to DB validator type checking.
+        raise TypeError(f"Got unexpected response type: {type(r)} {r}")
+      rows.append([r.client_id, r.flow_id, str(r.request_id), str(r.response_id),
+                   response,status,iterator,spanner_lib.COMMIT_TIMESTAMP])
+
+      txn.insert_or_update(table="FlowResponses", columns=columns, values=rows)
```

Comment:
```suggestion
    rows = []
    for r in responses:
      response = None
      status = None
      iterator = None
      if isinstance(r, flows_pb2.FlowResponse):
        response = r
      elif isinstance(r, flows_pb2.FlowStatus):
        status = r
      elif isinstance(r, flows_pb2.FlowIterator):
        iterator = r
      else:
        # This can't really happen due to DB validator type checking.
        raise TypeError(f"Got unexpected response type: {type(r)} {r}")
      rows.append([r.client_id, r.flow_id, str(r.request_id), str(r.response_id),
                   response,status,iterator,spanner_lib.COMMIT_TIMESTAMP])

    txn.insert_or_update(table="FlowResponses", columns=columns, values=rows)
```
Move the transaction outside of the loop